rules/ruleset.go

```diff
@@ -90,32 +110,149 @@ func newActiveRuleSet(
-	// TODO(xichen): pool the policies.
+// mappingPoliciesFor returns the mapping policies applicable to a given id at a given
+// time. It returns true if one or more matching policies are found, and false otherwise.
+func (as *activeRuleSet) mappingPoliciesFor(
+	id []byte,
+	isRollupID bool,
+	timeNanos int64,
+	opts MatchOptions,
+) (policy.StagedPolicies, bool) {
+	var (
+		sp      policy.StagedPolicies
+		matched bool
+	)
+	if !isRollupID {
+		sp, matched = as.mappingsForNonRollupID(id, timeNanos)
+	} else {
+		sp, matched = as.mappingsForRollupID(id, timeNanos)
+	}
+	if matched {
+		return sp, true
+	}
+	if opts.UseDefaultIfNoMatch {
+		return policy.DefaultStagedPolicies, true
+	}
+	return policy.DefaultStagedPolicies, false
+}
+
+// NB(xichen): in order to determine the applicable policies for a rollup metric, we need to
+// match the id against rollup rules to determine which rollup rules are applicable, under the
+// assumption that no two rollup targets in the same namespace may have the same rollup metric
+// name and the list of rollup tags. Otherwise, a rollup metric could potentially match more
+// than one rollup rule with different policies even though only one of the matched rules was
+// used to produce the given rollup metric id due to its tag filters, thereby causing the wrong
+// staged policies to be returned. This also implies at any given time, at most one rollup target
+// may match the given rollup id.
+func (as *activeRuleSet) mappingsForRollupID(id []byte, timeNanos int64) (policy.StagedPolicies, bool) {
+	if len(as.rollupRules) == 0 {
+		return policy.DefaultStagedPolicies, false
+	}
+
+	// If we cannot extract tags from the id, this is likely an invalid
+	// metric and we bail early.
+	name, tags, err := as.tagFilterOpts.NameAndTagsFn(id)
+	if err != nil {
+		return policy.DefaultStagedPolicies, false
+	}
+
+	var (
+		cutoverNanos int64
+		policies     []policy.Policy
+		matched      bool
+	)
+	for _, rollupRule := range as.rollupRules {
+		snapshot := rollupRule.ActiveSnapshot(timeNanos)
+		if snapshot == nil {
+			continue
+		}
+		for _, target := range snapshot.targets {
+			if !bytes.Equal(target.Name, name) {
+				continue
+			}
+			var (
+				tagIter      = as.tagFilterOpts.SortedTagIteratorFn(tags)
+				hasMoreTags  = tagIter.Next()
+				targetTagIdx = 0
+			)
+			for hasMoreTags && targetTagIdx < len(target.Tags) {
+				tagName, _ := tagIter.Current()
+				res := bytes.Compare(tagName, target.Tags[targetTagIdx])
+				if res == 0 {
+					targetTagIdx++
+					hasMoreTags = tagIter.Next()
+					continue
+				}
+				// If one of the target tags is not found in the id, this is considered
+				// a non-match so bail immediately.
+				if res > 0 {
+					break
+				}
+				hasMoreTags = tagIter.Next()
+			}
+			tagIter.Close()
+
+			// If all of the target tags are matched, this is considered as a match.
+			if targetTagIdx == len(target.Tags) {
+				matched = true
+				policies = append(policies, target.Policies...)
```

Comment:
We need to clone it so we don't end up mutating target policies when resolving them.